share/config.fish

```diff
@@ -195,6 +195,68 @@ function . --description 'Evaluate contents of file (deprecated, see "source")'
 	end
 end
 
+# Try to set the locale from the system configuration if we did not inherit any.
+# One case where this can happen is a linux with systemd, used via getty (and then possibly startx).
+# See https://github.com/fish-shell/fish-shell/issues/3092.
+# This isn't actually our job, so there's a bunch of edge-cases we are unlikely to handle properly.
+# If we get a value for _any_ language variable, we assume we've inherited something sensible to skip all this
+# and to allow the user to set it at runtime without mucking with config files.
+# NOTE: This breaks the expectation that an empty LANG will be the same as LANG=POSIX,
+# but an empty LANG seems more likely to be caused by mis- or missing configuration.
+
+# Do this before reading the snippets because they might be in UTF-8 (with non-ASCII characters).
+
+# Note the missing LC_ALL - locale.conf doesn't allow it, so we should not set it.
+set -l LANGVARS LANG LANGUAGE LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE LC_MONETARY LC_MESSAGES LC_PAPER LC_NAME LC_ADDRESS LC_TELEPHONE LC_MEASUREMENT LC_IDENTIFICATION
+
+# We use LC_ALL to figure out if we're without a locale, but we don't set it later.
+if not string length -q -- $$LANGVARS $LC_ALL
```

Comment:
Putting this in a separate, private, script also means you can rewrite this as 

```
if string length -q -- $$LANGVARS $LC_ALL
    return 0
end
```

Then the rest of the block doesn't need to be indented.